src/bblgum/impl.clj

```diff
@@ -30,3 +30,107 @@
                        str/trim
                        str/split-lines
                        (filter seq))))))
+
+(defn gum*
+  "Low level API. If you don't know why you are using `gum*` then you're probably looking for `gum`
+  Options map:
+  cmd: The interaction command. Can be a keyword. Required
+  opts: A map of options to be passed as optional params to gum
+  args: A vec of args to be passed as positional params to gum
+  in: An input stream than can be passed to gum
+  as: Coerce the output. Currently supports :bool, :ignored or defaults to a seq of strings
+  gum-path: Path to the gum binary. Defaults to gum
+  Returns a map of:
+  status: The exit code from gum
+  result: The output from the execution: seq of lines or coerced via :as."
+  [{:keys [cmd opts args in as gum-path]}]
+  (when-not cmd
+    (throw (IllegalArgumentException. ":cmd must be provided or non-nil")))
+  (let [gum-path (or gum-path "gum")
+        with-opts (->> opts
+                       (map (fn [[opt value]]
+                              (str "--" (i/->str opt) "=" (i/multi-opt value))))
+                       (into [gum-path (i/->str cmd)]))
+        out (if (= :ignored as) :inherit :string)
+        args (if (or (empty? args) (= "--" (first args)))
+               args
+               (cons "--" args))
+        {:keys [exit out]} (i/exec (into with-opts args) in out)]
+    {:status exit
+     :result (case as
+               :bool (zero? exit)
+               out)}))
+
+(defn prepare-options [m]
+  (let [fmted-keys [:opts :in :as :gum-path]
+        fmted (select-keys m fmted-keys)
+        opts (apply dissoc m fmted-keys)]
+    (update fmted :opts merge opts)))
+
+(defn prepare-cmd-map
+  "Prepares command map to be passed to `gum*`. Tries to be smart and figure out what user wants."
+  ([cmd]
+   (if (map? cmd)
+     cmd
+     (prepare-cmd-map cmd [] nil)))
+  ([cmd args-or-opts]
+   (if (vector? args-or-opts)
+     (prepare-cmd-map cmd args-or-opts nil)
+     (prepare-cmd-map cmd [] args-or-opts)))
+  ([cmd args & options]
+   (let [args* (if (vector? args) args [])
+         foptions (filter some? options)
+         options* (if (keyword? args) (conj foptions args) foptions)]
+     (merge {:cmd cmd :args args*} (if (seq options*) (prepare-options (apply hash-map options*)) {})))))
+
+(defn gum
```

Comment:
this can be in core tbh

Reply:
do see much of an upside to hide this one?

Reply:
We have a very clean public function without any implementation details for people to use. It shows arities, which helps somebody interested understand how calls work.

Everything else is in impl.clj, for those interested to see or submit prs, so all dev work can go on in one place. So I suggest leaving it as is.

Reply:
> dont see much of an upside to hide this one?

Explained below this question. You're splitting business logic in two files without a good reason. This way you have a public API in user facing ns, and all logic in private ns. Convenient for the user, convenient for us.

Reply:
Right, im just a bit fussed about the variety of `gum`s. 😓 
Maybe a rename like this?
- `i/gum*` -> `i/run`
- `i/gum` -> `i/dispatch`

Reply:
Sure, let me rename

Reply:
I would keep i/gum still called the same way, because it clearly communicates that the public `gum` is a front-end API for the private `gum`. That part is fine. I'll rename i/gum* to i/run though :)

Reply:
makes sense